File: consumer-fixture-langs/scala/src/test/scala/se/deversity/asynctest/fixture/GuardedCounterStaysSilentTest.scala
package se.deversity.asynctest.fixture

import org.junit.jupiter.api.{AfterAll, BeforeAll}
import se.deversity.asynctest.{AsyncFindings, AsyncTest, AsyncTestContext, FailOn}

/**
 * The other direction: the same write and hook inside `this.synchronized`. `AnyRef.synchronized`
 * is a monitorenter on `this`, so the detector's `Thread.holdsLock(owner)` probe sees the lock
 * and the write is not reported.
 */
object GuardedCounterStaysSilentTest {
  private var findings: AsyncFindings = scala.compiletime.uninitialized

  @BeforeAll
  def collect(): Unit =
    findings = AsyncFindings.collect()

  @AfterAll
  def noRaceWasReported(): Unit =
    try findings.assertNotReported("RaceConditionDetector")
    finally findings.close()
}

class GuardedCounterStaysSilentTest {

  private var counter = 0

  @AsyncTest(threads = 8, invocations = 200, failOn = FailOn.NONE, licenseMockMode = true)
  def guardedWritesFromEightThreads(): Unit = this.synchronized {
    val ctx = AsyncTestContext.get()
    if (ctx != null) ctx.sharedRaceConditionDetector().recordFieldWrite(this, "counter")
    counter += 1
  }
}

File: consumer-fixture-langs/clojure/src/test/clojure/se/deversity/asynctest/fixture/guarded_counter_stays_silent_test.clj
;; The other direction: the same write and hook inside (locking this ...). `locking` is a
;; monitorenter on the object, so the detector's Thread.holdsLock probe sees the lock and the
;; write is not reported. See lost_update_is_reported_test.clj for the gen-class notes.
(ns se.deversity.asynctest.fixture.guarded-counter-stays-silent-test
  (:import [se.deversity.asynctest AsyncFindings AsyncTestContext])
  (:gen-class
    :name se.deversity.asynctest.fixture.GuardedCounterStaysSilentTest
    :prefix "-"
    :methods [^{:static true} [^{org.junit.jupiter.api.BeforeAll {}}
                               collect [] void]
              ^{:static true} [^{org.junit.jupiter.api.AfterAll {}}
                               noRaceWasReported [] void]
              [^{se.deversity.asynctest.AsyncTest {:threads #=(int 8)
                                                    :invocations #=(int 200)
                                                    :failOn se.deversity.asynctest.FailOn/NONE
                                                    :licenseMockMode true}}
               guardedWritesFromEightThreads [] void]]))

(defonce ^:private findings (atom nil))
(defonce ^:private cell (long-array 1))

(defn -collect []
  (reset! findings (AsyncFindings/collect)))

(defn -noRaceWasReported []
  (let [^AsyncFindings f @findings]
    (try
      (.assertNotReported f "RaceConditionDetector")
      (finally
        (.close f)))))

(defn -guardedWritesFromEightThreads [this]
  (locking this
    (when-let [ctx (AsyncTestContext/get)]
      (.recordFieldWrite (.sharedRaceConditionDetector ctx) this "cell"))
    (let [^longs c cell]
      (aset c 0 (inc (aget c 0))))))
